rpc/src/eth_pubsub.rs

```diff
@@ -0,0 +1,146 @@
+use std::{marker::PhantomData, sync::Arc};
+use std::collections::BTreeMap;
+use sp_runtime::traits::{Block as BlockT, BlakeTwo256};
+use sp_transaction_pool::TransactionPool;
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_storage::StorageKey;
+use sp_io::hashing::twox_128;
+use sc_client_api::{
+	backend::{StorageProvider, Backend, StateBackend},
+	client::BlockchainEvents
+};
+use sc_rpc::Metadata;
+use log::warn;
+
+use jsonrpc_pubsub::{typed::Subscriber, SubscriptionId, manager::SubscriptionManager};
+use frontier_rpc_core::EthPubSubApi::{self as EthPubSubApiT};
+use frontier_rpc_core::types::{
+	Rich, Header, Bytes,
+	pubsub::{Kind, Params, Result as PubSubResult}
+};
+use ethereum_types::{H256, U256};
+use codec::Decode;
+
+pub use frontier_rpc_core::EthPubSubApiServer;
+use futures::{StreamExt as _, TryStreamExt as _};
+
+use jsonrpc_core::{Result as JsonRpcResult, futures::{Future, Sink}};
+
+pub struct EthPubSubApi<B: BlockT, P, C, BE> {
+	_pool: Arc<P>,
+	client: Arc<C>,
+	subscriptions: SubscriptionManager,
+	_marker: PhantomData<(B, BE)>,
+}
+
+impl<B: BlockT, P, C, BE> EthPubSubApi<B, P, C, BE> {
+	pub fn new(
+		_pool: Arc<P>,
+		client: Arc<C>,
+		subscriptions: SubscriptionManager,
+	) -> Self {
+		Self { _pool, client, subscriptions, _marker: PhantomData }
+	}
+}
+
+fn storage_prefix_build(module: &[u8], storage: &[u8]) -> Vec<u8> {
+	[twox_128(module), twox_128(storage)].concat().to_vec()
+}
+
+impl<B: BlockT, P, C, BE> EthPubSubApiT for EthPubSubApi<B, P, C, BE>
+	where
+		B: BlockT<Hash=H256> + Send + Sync + 'static,
+		P: TransactionPool<Block=B> + Send + Sync + 'static,
+		C: ProvideRuntimeApi<B> + StorageProvider<B,BE> +
+			BlockchainEvents<B> + HeaderBackend<B>,
+		C: Send + Sync + 'static,
+		BE: Backend<B> + 'static,
+		BE::State: StateBackend<BlakeTwo256>,
+{
+	type Metadata = Metadata;
+	fn subscribe(
+		&self,
+		_metadata: Self::Metadata,
+		subscriber: Subscriber<PubSubResult>,
+		kind: Kind,
+		_params: Option<Params>,
+	) {
+		match kind {
+			Kind::Logs => {
+				unimplemented!(); // TODO
+			},
+			Kind::NewHeads => {
+				let key: StorageKey = StorageKey(
+					storage_prefix_build(b"Ethereum", b"LatestHeader")
+				);
+				let stream = match self.client.storage_changes_notification_stream(
+					Some(&[key]),
+					None
+				) {
+					Ok(stream) => stream,
+					Err(_err) => {
+						unimplemented!(); // TODO
+					},
+				};
+				self.subscriptions.add(subscriber, |sink| {
+					let stream = stream
+						.map(|(_block, changes)| {
```

Comment:
Maybe move this to a separate function to reduce the indent.
this also apply to the inner: Header {....} part